In [7]:
#Importamos las bibliotecas a utilizar
import pandas as pd 
import numpy as np

url = 'https://raw.githubusercontent.com/MarioGHdz/FisicaComputacional1/master/Actividad%201/Datos%20diarios.txt'
# Definimos los nombres de las columnas
# El archivo tiene códigos ASCII extendido (Español), encoding='cp1251' Windows (Europa Occidental)
# El separador de datos es de uno o más espacios, sep='\s+'
# Se brincará los primeros 19 renglones y el último, afirmamos que no nay encabezado
# Se procesan los datos utilizando Python, engine='python' 
my_cols = ['Fecha', 'Precip', 'Evap', 'Tmax', 'Tmin']
df_dist = pd.read_csv(url, names=my_cols, encoding='cp1251', sep='\s+', header=None, skiprows=19, skipfooter=1, engine='python')

# Revisamos el encabezado y el final del dataframe
print(df_dist.head(10))
print(df_dist.tail(10))




        Fecha Precip Evap  Tmax  Tmin
0  01/05/1969      0    9  30.5   8.5
1  02/05/1969      0  9.1  30.5   8.5
2  03/05/1969      0    9    30     9
3  04/05/1969      0  7.1  27.5    12
4  05/05/1969      5  4.3    28   7.5
5  06/05/1969   13.3  5.1  30.5     6
6  07/05/1969    2.2  4.2    29     5
7  08/05/1969      0  5.9    24   8.5
8  09/05/1969      0    9    30  11.5
9  10/05/1969      0  7.2  34.5    14
            Fecha Precip  Evap Tmax Tmin
16489  21/04/2016      0  Nulo   33    9
16490  22/04/2016      0  Nulo   33   10
16491  23/04/2016      0  Nulo   30   12
16492  24/04/2016      0  Nulo   30   10
16493  25/04/2016      0  Nulo   27   12
16494  26/04/2016      0  Nulo   29    8
16495  27/04/2016      0  Nulo   28    7
16496  28/04/2016      0  Nulo   27    7
16497  29/04/2016      0  Nulo   25    5
16498  30/04/2016      0  Nulo   26    7


In [4]:

# Investigamos su tamaño con la función shape (renglones, columnas)
df_dist.shape

(16499, 5)

In [10]:
# Información del dataframe
df_dist.info()


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16499 entries, 0 to 16498
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Fecha   16499 non-null  object
 1   Precip  16499 non-null  object
 2   Evap    16499 non-null  object
 3   Tmax    16499 non-null  object
 4   Tmin    16499 non-null  object
dtypes: object(5)
memory usage: 644.6+ KB


In [8]:

# Hacemos una copia de trabajo y preservamos el dataframe original
df_work0 = df_dist.copy()
df_work0.head()

Fecha Precip Evap  Tmax Tmin
0  01/05/1969      0    9  30.5  8.5
1  02/05/1969      0  9.1  30.5  8.5
2  03/05/1969      0    9    30    9
3  04/05/1969      0  7.1  27.5   12
4  05/05/1969      5  4.3    28  7.5

In [11]:
df_work0 = df_dist.copy()
df_work0.tail()

Fecha Precip  Evap Tmax Tmin
16494  26/04/2016      0  Nulo   29    8
16495  27/04/2016      0  Nulo   28    7
16496  28/04/2016      0  Nulo   27    7
16497  29/04/2016      0  Nulo   25    5
16498  30/04/2016      0  Nulo   26    7

In [9]:
# Reemplazamos los textos 'Nulo' por espacio en blanco ''
str_Nulo = 'Nulo'
df_work1 = df_work0.replace(to_replace=str_Nulo, value='', regex=True)
df_work1.head()

Fecha Precip Evap  Tmax Tmin
0  01/05/1969      0    9  30.5  8.5
1  02/05/1969      0  9.1  30.5  8.5
2  03/05/1969      0    9    30    9
3  04/05/1969      0  7.1  27.5   12
4  05/05/1969      5  4.3    28  7.5

In [16]:
str_Nulo = 'Nulo'
df_work1 = df_work0.replace(to_replace=str_Nulo, value='', regex=True)
df_work1.tail()

Fecha Precip Evap Tmax Tmin
16494  26/04/2016      0        29    8
16495  27/04/2016      0        28    7
16496  28/04/2016      0        27    7
16497  29/04/2016      0        25    5
16498  30/04/2016      0        26    7

In [13]:
# Imprimimos la información actualizada del dataframe
df_work1.info()
print(df_work1.head())


<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16499 entries, 0 to 16498
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Fecha   16499 non-null  object
 1   Precip  16499 non-null  object
 2   Evap    16499 non-null  object
 3   Tmax    16499 non-null  object
 4   Tmin    16499 non-null  object
dtypes: object(5)
memory usage: 644.6+ KB
        Fecha Precip Evap  Tmax Tmin
0  01/05/1969      0    9  30.5  8.5
1  02/05/1969      0  9.1  30.5  8.5
2  03/05/1969      0    9    30    9
3  04/05/1969      0  7.1  27.5   12
4  05/05/1969      5  4.3    28  7.5


In [15]:
df_work1.info()
print(df_work1.tail())

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16499 entries, 0 to 16498
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype 
---  ------  --------------  ----- 
 0   Fecha   16499 non-null  object
 1   Precip  16499 non-null  object
 2   Evap    16499 non-null  object
 3   Tmax    16499 non-null  object
 4   Tmin    16499 non-null  object
dtypes: object(5)
memory usage: 644.6+ KB
            Fecha Precip Evap Tmax Tmin
16494  26/04/2016      0        29    8
16495  27/04/2016      0        28    7
16496  28/04/2016      0        27    7
16497  29/04/2016      0        25    5
16498  30/04/2016      0        26    7


In [17]:
# Deseamos convertir los valores de las variables ambientales a núm. flotantes 
cols_list = ['Precip', 'Evap', 'Tmax', 'Tmin']
for cols in cols_list:
  df_work1[cols] = pd.to_numeric(df_work1[cols], errors='coerce')

df_work1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16499 entries, 0 to 16498
Data columns (total 5 columns):
 #   Column  Non-Null Count  Dtype  
---  ------  --------------  -----  
 0   Fecha   16499 non-null  object 
 1   Precip  16498 non-null  float64
 2   Evap    11586 non-null  float64
 3   Tmax    16465 non-null  float64
 4   Tmin    16467 non-null  float64
dtypes: float64(4), object(1)
memory usage: 644.6+ KB


In [18]:
# De la instrucción anterior, vemos que hay valores faltantes. Sumamos los valores nulos. 
df_work1.isnull().sum()

Fecha        0
Precip       1
Evap      4913
Tmax        34
Tmin        32
dtype: int64

In [19]:
print(df_work1.head())
print(df_work1.head())

        Fecha  Precip  Evap  Tmax  Tmin
0  01/05/1969     0.0   9.0  30.5   8.5
1  02/05/1969     0.0   9.1  30.5   8.5
2  03/05/1969     0.0   9.0  30.0   9.0
3  04/05/1969     0.0   7.1  27.5  12.0
4  05/05/1969     5.0   4.3  28.0   7.5
        Fecha  Precip  Evap  Tmax  Tmin
0  01/05/1969     0.0   9.0  30.5   8.5
1  02/05/1969     0.0   9.1  30.5   8.5
2  03/05/1969     0.0   9.0  30.0   9.0
3  04/05/1969     0.0   7.1  27.5  12.0
4  05/05/1969     5.0   4.3  28.0   7.5


In [20]:
print(df_work1.tail())
print(df_work1.tail())

            Fecha  Precip  Evap  Tmax  Tmin
16494  26/04/2016     0.0   NaN  29.0   8.0
16495  27/04/2016     0.0   NaN  28.0   7.0
16496  28/04/2016     0.0   NaN  27.0   7.0
16497  29/04/2016     0.0   NaN  25.0   5.0
16498  30/04/2016     0.0   NaN  26.0   7.0
            Fecha  Precip  Evap  Tmax  Tmin
16494  26/04/2016     0.0   NaN  29.0   8.0
16495  27/04/2016     0.0   NaN  28.0   7.0
16496  28/04/2016     0.0   NaN  27.0   7.0
16497  29/04/2016     0.0   NaN  25.0   5.0
16498  30/04/2016     0.0   NaN  26.0   7.0


In [21]:
# Un análisis estadístico breve de las variables numéricas redondeado a 2 cifras significativas
df_work1.describe().round(2)

Precip      Evap      Tmax      Tmin
count  16498.00  11586.00  16465.00  16467.00
mean       1.23      6.03     28.62      9.88
std        5.46      3.20      7.25      7.17
min        0.00      0.10      1.50    -10.50
25%        0.00      3.50     23.50      4.00
50%        0.00      5.70     29.50      8.50
75%        0.00      8.20     34.50     16.50
max      160.00     17.00     46.00     26.00

No se noto nada fuera de lo normal, lo único que se puede destacar que la temperatura mínima alcanzó el máximo de -10.50°.

In [23]:
# Volvemos a hacer una pausa y definimos un nuevo dataframe por si hay que regresarse.
df_work2 = df_work1.copy()
df_work2['Fecha'].head()

0    01/05/1969
1    02/05/1969
2    03/05/1969
3    04/05/1969
4    05/05/1969
Name: Fecha, dtype: object

In [24]:
# Pandas tiene la función pd.to_datetime que convierte una cadena en tipo datetime64[ns]
# Intentamos a ver si Python reconoce el formato en nuestro caso. 
# Agregamos la opción de que el primer dato es el día (dayfirst=True)
df_work2['Fecha']=pd.to_datetime(df_work2['Fecha'], dayfirst=True).copy()
print(df_work1.head())

        Fecha  Precip  Evap  Tmax  Tmin
0  01/05/1969     0.0   9.0  30.5   8.5
1  02/05/1969     0.0   9.1  30.5   8.5
2  03/05/1969     0.0   9.0  30.0   9.0
3  04/05/1969     0.0   7.1  27.5  12.0
4  05/05/1969     5.0   4.3  28.0   7.5


In [25]:
# Verificamos que cada variable (columna), sea del tipo que esperamos.
# Observamos que la variable Fecha es de tipo 'datetime64[ns]'  
df_work2.dtypes

Fecha     datetime64[ns]
Precip           float64
Evap             float64
Tmax             float64
Tmin             float64
dtype: object

In [26]:
# Buscamos ahora añadir dos columnas adicionales: df['Año'], df['Mes']
# Usaremos la función .dt.year y .dt.month para extraer el Año y Mes de la Fecha
df_work2['Año'] = df_work2['Fecha'].dt.year
df_work2['Mes'] = df_work2['Fecha'].dt.month

In [27]:
# Verificamos que se hayan creado de forma correcta
# Tendremos ahora dos columnas adicionales que podemos utilizar para filtrar información.
print(df_work2.head())
print(df_work2.tail())

       Fecha  Precip  Evap  Tmax  Tmin   Año  Mes
0 1969-05-01     0.0   9.0  30.5   8.5  1969    5
1 1969-05-02     0.0   9.1  30.5   8.5  1969    5
2 1969-05-03     0.0   9.0  30.0   9.0  1969    5
3 1969-05-04     0.0   7.1  27.5  12.0  1969    5
4 1969-05-05     5.0   4.3  28.0   7.5  1969    5
           Fecha  Precip  Evap  Tmax  Tmin   Año  Mes
16494 2016-04-26     0.0   NaN  29.0   8.0  2016    4
16495 2016-04-27     0.0   NaN  28.0   7.0  2016    4
16496 2016-04-28     0.0   NaN  27.0   7.0  2016    4
16497 2016-04-29     0.0   NaN  25.0   5.0  2016    4
16498 2016-04-30     0.0   NaN  26.0   7.0  2016    4


In [28]:
# Verificamos que Año y Mes son variables enteras.

df_work2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16499 entries, 0 to 16498
Data columns (total 7 columns):
 #   Column  Non-Null Count  Dtype         
---  ------  --------------  -----         
 0   Fecha   16499 non-null  datetime64[ns]
 1   Precip  16498 non-null  float64       
 2   Evap    11586 non-null  float64       
 3   Tmax    16465 non-null  float64       
 4   Tmin    16467 non-null  float64       
 5   Año     16499 non-null  int64         
 6   Mes     16499 non-null  int64         
dtypes: datetime64[ns](1), float64(4), int64(2)
memory usage: 902.4 KB
